File: src/main/java/chapter4/transporm/ScanExample.java
package chapter4.transporm;

import common.Log;
import io.reactivex.Observable;

public class ScanExample {

    public void marbleDiagram() {
        String[] balls = {"1","3","5"};
        Observable<String> source = Observable.fromArray(balls)
                .scan((ball1, ball2) -> ball2 + "(" + ball1 + ")");
        source.subscribe(Log::it);
    }

    public static void main(String[] args) {
        ScanExample demo = new ScanExample();
        demo.marbleDiagram();
    }

}

File: src/main/kotlin/chapter4/transporm/ScanExample_.kt
package chapter4.transporm

import common.Log_
import io.reactivex.Observable

class ScanExample_ {

    fun marbleDiagram() {
        val balls = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*balls)
                .scan { ball1, ball2 -> "$ball2($ball1)" }
        source.subscribe { Log_.it(it) }
    }

}


fun main(args: Array<String>) {
    val demo = ScanExample_()
    demo.marbleDiagram()
}
